### PR TITLE
Enable loiter type offboard position setpoint for Fixedwing vehicles

### LINKED ISSUE #13180: Offboard position setpoints with loiter type mask makes Fixed wing vehicle fly away
**Describe the bug**
Offboard position setpoints with loiter type masks result in a fly away of a fixed wing vehicle. The roll and pitch setpoints are set to zero and continues to fly to the last direction where offboard mode was triggered. The offboard position setpoints fly as expected without the loiter type mask.

Screenshot from 2019-10-14 12-05-00

**To Reproduce**
1. Take off FW vehicle in SITL
2. Send `SET_POSITION_LOCAL_NED` setpoint with `type_mask=0x3000` ([see](https://github.com/PX4/Firmware/blob/master/src/modules/mavlink/mavlink_receiver.cpp#L774)
3. Swich mode to offboard mode

**Expected behavior**
The FW vehicle circles around the loiter setpoint

**Log Files and Screenshots**
The [log](https://review.px4.io/plot_app?log=8617fe45-3269-4098-9165-e94b176d74a4) shows the offboard setpoint resulting in a fly away. A local position setpoint of [0.0, 0.0, 20.0] was being streemed with`type_mask=0x3000`

**Additional context**
Currently flying offboard position setpoints are passed as a reference position for the L1 controller, which results the vehicle turning around and traversing the setpoint as can be seen [here](https://review.px4.io/plot_app?log=209a8f30-a2aa-45f1-85b0-6755170b88a2). Supporting circling the waypoints may be useful for various robotic applications using fixed wing.

### PR DESCRIPTION
**Pull request by Jaeyoung-Lim**
**Describe problem solved by this pull request**
This PR Fixes #13180 , where sending the loiter type offboard position setpoint on a fixed wing vehicle will result in a fly away. Loiter setpoints can be useful for fixed wing offboard mode as normal position setpoints are passed as a reference to the l1 controller so there are no explicit constraints on the radius of the vehicle while keeping close to the position setpoint.

The reason the it wasn't working was 
- Commander thinks it is a valid setpoint, as it is receiving enough `SET_POSITION_TARGET_LOCAL_NED` messages, therefore the vehicle is able to switch to offboard mode.
- The fixed wing position controller treats the setpoint as a loiter setpoint as it is marked [here](https://github.com/PX4/Firmware/blob/master/src/modules/mavlink/mavlink_receiver.cpp#L805) as a lotier setpoint. 
- However, as the `loiter_radius` and `loiter_direction` is not set from `SET_POSITION_TARGET_LOCAL_NED`, it is left as zero.

**Describe your solution**
Set the `loiter_radius` and `lotier_direction` if  `loiter_radius<=0`it is invalid (which probabiliy means it is in offboard mode following position setpoints). It uses the default `NAV_LOTIER_RAD` parameter as default loiter radius

**Test data / coverage**
This [log](https://review.px4.io/plot_app?log=314330b4-5013-4e91-8a5f-47c923e2a715) shows a flight which is being sent a offboard local position setpoint at [0.0m, 0.0m, 20.0m] and being switched between mission mode and offboard mode

### CODE REVIEW

**Comment by Jaeyoung-Lim** [comment]
@hamishwillee The types of setpoints are in general `UNDOCUMENTED`. I realized this is a thing by going through the code. Would it make sense to add this to the documentation?

**Comment by hamishwillee** [comment]
> The types of position setpoints are in general UNDOCUMENTED. I realized this is a thing by going through the code. Now that it works would it make sense to add this to the documentation?

@Jaeyoung-Lim 
How does the "type" of the setpoint get determined? I.e. I can't see it in https://mavlink.io/en/messages/common.html#SET_POSITION_TARGET_LOCAL_NED so it looks entirely internal to PX4. 
Can someone determine what type of setpoint they are dealing with?

Upshot, I can't answer this yet, because I don't understand where these come from and how they are used.

**Comment by Jaeyoung-Lim** [comment]
@hamishwillee Exactly. It seems to be coming from https://github.com/PX4/Firmware/pull/3405 and the type definitions didn't make it through the mavlink spec. I do like the fact that we can use the `type_mask` to configure how the setpoint can be followed. But in my opinion, this should either be pushed to the mavlink standard or at least be documented somewhere.

**Comment by hamishwillee** [comment]
@Jaeyoung-Lim Thanks for the clarification. I don't know if this should be public or not, but if it is then IMO it has been done incorrectly.  

I have created https://github.com/mavlink/mavlink/pull/1250 to discuss whether it should be public or not. Depending on how that falls out will define how this is documented. 

Can you tell me the difference between the idle and loiter setpoints, and why knowing the setpoint type is useful (as an answer to the linked PR?)

**Comment by Jaeyoung-Lim** [comment]
@hamishwillee Specifying the position of the setpoint does not always fully define the expected behavior of the vehicle.

An idle setpoint will make the vehicle be in idle state - zero throttle, zero roll / pitch. 
A Loiter setpoint will make the vehicle circle around the setpoint.
A normally passed postion setpoint will make the vehicle use the position setpoint as a reference to the L1 controller, so it will fly in a flower like pattern. 

All three types use the same position setpoint, but the behavior is handled differently in the fixed wing position controller.

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 1]
```diff
@@ -868,8 +870,17 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
 		} else if (pos_sp_curr.type == position_setpoint_s::SETPOINT_TYPE_LOITER) {
 
 			/* waypoint is a loiter waypoint */
-			_l1_control.navigate_loiter(curr_wp, curr_pos, pos_sp_curr.loiter_radius,
-						    pos_sp_curr.loiter_direction, nav_speed_2d);
+			float loiter_radius = pos_sp_curr.loiter_radius;
+			uint8_t loiter_direction = pos_sp_curr.loiter_direction;
+
+			if (pos_sp_curr.loiter_radius <= 0) { //TODO: Get loiter radius from parameter
```
Comment (sfuhrer) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: You now could remove the TODO, no?

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp [hunk 2]
```diff
@@ -868,8 +870,17 @@ FixedwingPositionControl::control_position(const Vector2f &curr_pos, const Vecto
...
+				loiter_radius = _parameters.loiter_radius;
+				loiter_direction = 1;
```
Comment (sfuhrer) [inline src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: You could maybe have the direction be set depending on the sign of the radius, similar to [here](https://github.com/PX4/Firmware/blob/cf8f03f190d787d19225704b4badc508dcf4797e/src/modules/navigator/mission_block.cpp#L526).

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 1 above)
Comment (Jaeyoung-Lim) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: Yes!

File: src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp (hunk 2 above)
Comment (Jaeyoung-Lim) [reply src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp]: Thanks, was not aware having lotier radius as a negative value was possible. Addressed on last commit

**Comment by hamishwillee** [comment]
@Jaeyoung-Lim What do the setpoints do on other vehicle types?

**Comment by Jaeyoung-Lim** [comment]
@hamishwillee I have replied in the user guide!

**Comment by Jaeyoung-Lim** [comment]
@sfuhrer Can we get this merged?

**Comment by sfuhrer** [comment]
> @sfuhrer Can we get this merged?

Yes!

**Comment by dagar** [comment]
This really makes me think that loiter radius should be a FW parameter rather than navigator.

**Pull request merged**